![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Fworkflows%2Fchurn_retention&file=churn_retention.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Fworkflows%2Fchurn_retention%2Fchurn_retention.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/workflows/churn_retention/churn_retention.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Churn / Retention — BigQuery ML

Define churn from **real order-history gaps** — not a synthetic label — then engineer features, train a classifier, evaluate it honestly, and identify the real drivers. Reuses [`workflows/customer_segmentation/`](../customer_segmentation/)'s dataset for narrative continuity: same customers, same time window, a related business question.

**Models used:** `BOOSTED_TREE_CLASSIFIER`
**Functions used:** `ML.EVALUATE`, `ML.CONFUSION_MATRIX`, `ML.GLOBAL_EXPLAIN`, `ML.FEATURE_IMPORTANCE`, `ML.EXPLAIN_PREDICT`

`BOOSTED_TREE_CLASSIFIER`'s own mechanics (confusion matrix, ROC curve, `TRANSFORM`, exporting/visualizing a tree) are already covered in depth in [`models/boosted_tree_classifier/`](../../models/boosted_tree_classifier/) and not repeated here — **this workflow's real content is the label definition, feature engineering, and honest evaluation.**

**Data:** [`bigquery-public-data.thelook_ecommerce`](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets) (`orders`, `order_items`, `users`).

**References:** [Full reference](../../RESOURCES.md) | [CREATE MODEL (BOOSTED_TREE_CLASSIFIER) docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree) | [Setup guide](../../setup/)

---
## Setup

Set your project and location, authenticate, and create a shared dataset. No connection needed.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics', 'bigframes')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready


---
## Step 1 — Define churn from a fixed feature cutoff + forward label window

**Feature cutoff:** `2024-01-01` (matches [`customer_segmentation`](../customer_segmentation/)). **Label window:** the following 180 days (through `2024-06-29`). A customer who ordered *before* the cutoff but placed **no order in the label window** is labeled `churned = TRUE`.

Both dates are far enough in the past (this notebook is being run well after `2024-06-29`) that this is **real observed history**, not a simulated forward projection — a genuine backtest, not a synthetic label.

In [5]:
query = """
CREATE OR REPLACE TABLE `{project}.{dataset}.churn_retention_features` AS
WITH order_value AS (
  SELECT order_id, user_id, SUM(sale_price) AS order_total,
         COUNT(DISTINCT product_id) AS n_products,
         MAX(CASE WHEN returned_at IS NOT NULL THEN 1 ELSE 0 END) AS had_return
  FROM `bigquery-public-data.thelook_ecommerce.order_items`
  GROUP BY order_id, user_id
),
rfm AS (
  SELECT
    o.user_id,
    DATE_DIFF(DATE '2024-01-01', MAX(DATE(o.created_at)), DAY) AS recency_days,
    COUNT(DISTINCT o.order_id) AS frequency,
    SUM(ov.order_total) AS monetary,
    SUM(ov.order_total) / COUNT(DISTINCT o.order_id) AS avg_order_value,
    SUM(ov.n_products) AS distinct_products,
    AVG(ov.had_return) AS return_rate
  FROM `bigquery-public-data.thelook_ecommerce.orders` o
  JOIN order_value ov USING (order_id, user_id)
  WHERE o.created_at < '2024-01-01'
  GROUP BY o.user_id
),
future_activity AS (
  SELECT DISTINCT user_id
  FROM `bigquery-public-data.thelook_ecommerce.orders`
  WHERE created_at >= '2024-01-01' AND created_at < '2024-06-29'
)
SELECT
  r.*, u.age, u.gender, u.traffic_source,
  DATE_DIFF(DATE '2024-01-01', DATE(u.created_at), DAY) AS tenure_days,
  fa.user_id IS NULL AS churned
FROM rfm r
JOIN `bigquery-public-data.thelook_ecommerce.users` u ON r.user_id = u.id
LEFT JOIN future_activity fa USING (user_id)
""".format(project=PROJECT_ID, dataset=DATASET_ID)
client.query(query).result()

query = f"""
SELECT
  COUNTIF(churned) AS n_churned,
  COUNTIF(NOT churned) AS n_retained,
  ROUND(AVG(CAST(churned AS INT64)), 3) AS churn_rate
FROM `{PROJECT_ID}.{DATASET_ID}.churn_retention_features`
"""
client.query(query).to_dataframe()

,n_churned,n_retained,churn_rate
0,25280,2436,0.912


**91.2% of customers churned.** The extreme imbalance is itself a real finding for this dataset: most customers never place a second order at all, regardless of any feature — consistent with the ~1.3 orders/customer average already seen in `customer_segmentation`.

---
## Step 2 — Baseline: `BOOSTED_TREE_CLASSIFIER` on RFM features alone

With a 91% base churn rate, a model that always predicts "churned" gets 91% accuracy while being completely useless. **`auto_class_weights = TRUE`** forces the model to actually learn the minority (retained) class instead of exploiting the imbalance.

In [6]:
query = """
CREATE OR REPLACE MODEL `{project}.{dataset}.churn_retention_baseline`
OPTIONS(
  model_type = 'BOOSTED_TREE_CLASSIFIER',
  input_label_cols = ['churned'],
  auto_class_weights = TRUE,
  data_split_method = 'AUTO_SPLIT',
  enable_global_explain = TRUE
) AS
SELECT recency_days, frequency, monetary, churned
FROM `{project}.{dataset}.churn_retention_features`
""".format(project=PROJECT_ID, dataset=DATASET_ID)
client.query(query).result()
print('Model churn_retention_baseline created')

Model churn_retention_baseline created


In [7]:
query = f"SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_baseline`)"
client.query(query).to_dataframe()

,precision,recall,accuracy,f1_score,log_loss,roc_auc
0,0.921723,0.484735,0.493274,0.635343,0.680201,0.531122


**`roc_auc` ≈ 0.53 — barely above random (0.5).** Recency/frequency/monetary from a single snapshot do not meaningfully separate churners from retained customers in this dataset. Precision looks deceptively high (~0.92) only because the model is right whenever it (correctly, and often) predicts the majority class.

---
## Step 3 — Richer feature engineering: behavior + demographics

Add average order value, distinct products purchased, return rate (from `order_items`), and account age plus demographics (from `users`) — signals a bare RFM table can't see.

In [8]:
query = """
CREATE OR REPLACE MODEL `{project}.{dataset}.churn_retention_model`
OPTIONS(
  model_type = 'BOOSTED_TREE_CLASSIFIER',
  input_label_cols = ['churned'],
  auto_class_weights = TRUE,
  data_split_method = 'AUTO_SPLIT',
  enable_global_explain = TRUE
) AS
SELECT recency_days, frequency, monetary, avg_order_value, distinct_products,
       return_rate, age, gender, traffic_source, tenure_days, churned
FROM `{project}.{dataset}.churn_retention_features`
""".format(project=PROJECT_ID, dataset=DATASET_ID)
client.query(query).result()
print('Model churn_retention_model created')

Model churn_retention_model created


In [9]:
query = f"SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`)"
client.query(query).to_dataframe()

,precision,recall,accuracy,f1_score,log_loss,roc_auc
0,0.92894,0.605024,0.596564,0.732782,0.672963,0.544204


**Honest finding, worth stating plainly rather than glossing over:** accuracy (0.49 → 0.60), recall (0.48 → 0.61), and F1 (0.64 → 0.73) all improve meaningfully with the richer feature set — but **`roc_auc` barely moves (0.531 → 0.544)**. This isn't a contradiction: `roc_auc` measures the model's *ranking* quality across every possible probability threshold, while accuracy/recall/F1 are all read off a single fixed 0.5 cutoff. Richer features can shift *where* predicted probabilities land relative to 0.5 — improving the fixed-threshold metrics — without meaningfully improving how well the model *ranks* churners above non-churners overall, which is what `roc_auc` captures. No single metric tells the whole story here. The practical takeaway: this synthetic e-commerce generator simply does not encode a strong *individual-level* churn signal in transaction history alone. A real production churn model would typically need richer behavioral signal — browsing/session data, email engagement, support tickets — than a transaction log by itself provides.

`ML.CONFUSION_MATRIX` on the richer model shows exactly how that 0.605 recall breaks down: of 5,056 real churners, it correctly flags 3,059 (the `expected_label=true, TRUE` cell) and misses 1,997 — recall on the **churned** class (`TRUE`), matching `ML.EVALUATE`'s figure exactly. The 234 false positives on the retained class are the real cost of that improved catch rate.

In [10]:
query = f"SELECT * FROM ML.CONFUSION_MATRIX(MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`)"
client.query(query).to_dataframe()

,expected_label,FALSE,TRUE
0,false,240,234
1,true,1997,3059


---
## Step 4 — `ML.GLOBAL_EXPLAIN` + `ML.FEATURE_IMPORTANCE`: what actually drives the model

`ML.GLOBAL_EXPLAIN` requires `enable_global_explain = TRUE` at training time (set above). Full mechanics of both functions are covered in [`models/boosted_tree_classifier/`](../../models/boosted_tree_classifier/) — here we just read the result.

In [11]:
query = f"SELECT * FROM ML.GLOBAL_EXPLAIN(MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`) ORDER BY attribution DESC"
client.query(query).to_dataframe()

,feature,attribution
0,tenure_days,0.072060
1,frequency,0.028365
2,recency_days,0.015414
3,age,0.014732
4,traffic_source,0.009623
5,avg_order_value,0.007973
6,monetary,0.005881
7,return_rate,0.005471
8,distinct_products,0.003170
9,gender,0.000000


**`tenure_days` (how long someone has been a customer) is the single dominant driver** — more important than any of the three original RFM features. `frequency` is second.

In [12]:
query = f"SELECT * FROM ML.FEATURE_IMPORTANCE(MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`) ORDER BY importance_gain DESC"
client.query(query).to_dataframe()

,feature,importance_weight,importance_gain,importance_cover
0,tenure_days,28,33.492179,3570.587289
1,frequency,16,18.579082,2857.387360
2,return_rate,4,14.496144,1180.438157
3,age,36,8.217875,990.515110
4,recency_days,42,8.035588,690.308041
5,distinct_products,6,6.857787,1056.706035
6,monetary,18,6.789050,746.451096
7,traffic_source,12,6.457027,1190.775674
8,avg_order_value,28,6.366187,553.210222
9,gender,0,0.000000,0.000000


`ML.FEATURE_IMPORTANCE`'s ranking (by `importance_gain`, the average loss reduction the feature contributes when it's used to split) doesn't match `ML.GLOBAL_EXPLAIN`'s ranking (by Shapley-style `attribution`) — `return_rate` jumps from 8th-of-9 by attribution to 3rd by gain, while `recency_days` drops from 3rd to 5th. These measure genuinely different things: `importance_gain` reflects how useful a feature is *when the tree chooses to split on it* (a feature used rarely but decisively, like `return_rate` at only `importance_weight=4`, can still have high average gain), while `attribution` reflects the feature's typical contribution *across all predictions*, split frequency included. Neither ranking is "more correct" — they answer different questions ("how much does this feature matter each time it's used?" vs. "how much does this feature matter on average across customers?").

---
## Step 5 — `ML.EXPLAIN_PREDICT`: per-customer driver attribution

Global importance says what matters *on average*; `ML.EXPLAIN_PREDICT` says what mattered for *this specific customer* — the level of detail an actual retention campaign would act on.

In [13]:
query = f"""
SELECT
  predicted_churned,
  top_feature_attributions
FROM ML.EXPLAIN_PREDICT(
  MODEL `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`,
  (SELECT * FROM `{PROJECT_ID}.{DATASET_ID}.churn_retention_features` LIMIT 5),
  STRUCT(3 AS top_k_features)
)
"""
client.query(query).to_dataframe()

,predicted_churned,top_feature_attributions
0,True,"[{'feature': 'tenure_days', 'attribution': 0.09380176663398743}, {'feature': 'frequency', 'attribution': 0.029024917632341385}, {'feature': 'age', 'attribution': 0.004933916963636875}]"
1,True,"[{'feature': 'tenure_days', 'attribution': 0.09486702084541321}, {'feature': 'frequency', 'attribution': 0.016664814203977585}, {'feature': 'recency_days', 'attribution': 0.01454149279743433}]"
2,True,"[{'feature': 'recency_days', 'attribution': 0.07149603217840195}, {'feature': 'tenure_days', 'attribution': 0.05478781834244728}, {'feature': 'frequency', 'attribution': 0.028314122930169106}]"
3,False,"[{'feature': 'tenure_days', 'attribution': 0.08887118101119995}, {'feature': 'recency_days', 'attribution': -0.011274799704551697}, {'feature': 'frequency', 'attribution': -0.010414297692477703}]"
4,True,"[{'feature': 'age', 'attribution': 0.13093560934066772}, {'feature': 'tenure_days', 'attribution': 0.1147245392203331}, {'feature': 'monetary', 'attribution': -0.04213528707623482}]"


---
## Step 6 — Tie back to `customer_segmentation`: does loyalty actually reduce churn risk?

[`customer_segmentation`](../customer_segmentation/)'s KMEANS segments are driven jointly by recency, frequency, *and* monetary value — there's no single hard cutoff separating them. As a simple proxy for its "loyal regulars"/"champions" segments (both averaging 2+ orders) vs. its "recent, low-value" and "lapsed" segments (both averaging 1 order), split customers by order count alone: does 2+ orders actually predict lower churn?

In [14]:
query = f"""
SELECT
  CASE WHEN frequency = 1 THEN '1 order' ELSE '2+ orders' END AS frequency_group,
  COUNT(*) AS n_customers,
  ROUND(AVG(CAST(churned AS INT64)), 3) AS churn_rate
FROM `{PROJECT_ID}.{DATASET_ID}.churn_retention_features`
GROUP BY frequency_group
ORDER BY frequency_group
"""
client.query(query).to_dataframe()

,frequency_group,n_customers,churn_rate
0,1 order,21558,0.916
1,2+ orders,6158,0.900


**A genuinely useful, slightly deflating finding:** 1-order customers churn at 91.6%; repeat customers (2+ orders) still churn at 90.0% — barely lower. In this dataset, having already placed a second order does **not** meaningfully protect against churn. This helps explain why the classifier's `roc_auc` stays low even with richer features: the signal genuinely is weak, not just poorly engineered.

---
## Related content

- [`models/boosted_tree_classifier/`](../../models/boosted_tree_classifier/) — the algorithm mechanics in depth: `ML.CONFUSION_MATRIX`, `ML.ROC_CURVE`, exporting/visualizing a tree, the `TRANSFORM` clause.
- [`workflows/customer_segmentation/`](../customer_segmentation/) — the shared dataset and RFM feature engineering this workflow builds on.

---
## Examples — `%%bigquery` Magics

The same operations using IPython magic commands — write SQL directly in cells without Python string wrapping.

In [15]:
%%bigquery --project {PROJECT_ID}

SELECT * FROM ML.GLOBAL_EXPLAIN(MODEL `statmike-mlops-349915.bq_ml.churn_retention_model`)
ORDER BY attribution DESC

Query is running:   0%|          |

Downloading:   0%|          |

,feature,attribution
0,tenure_days,0.072060
1,frequency,0.028365
2,recency_days,0.015414
3,age,0.014732
4,traffic_source,0.009623
5,avg_order_value,0.007973
6,monetary,0.005881
7,return_rate,0.005471
8,distinct_products,0.003170
9,gender,0.000000


---
## Examples — BigFrames

`bigframes.ml.ensemble.XGBClassifier` maps to `BOOSTED_TREE_CLASSIFIER` (already verified in `models/boosted_tree_classifier/`).

In [16]:
import bigframes.pandas as bpd
from bigframes.ml.ensemble import XGBClassifier

bpd.close_session()  # Reset session to apply project/location settings
bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

df = bpd.read_gbq(f"SELECT recency_days, frequency, monetary, avg_order_value, distinct_products, return_rate, age, gender, traffic_source, tenure_days, churned FROM `{PROJECT_ID}.{DATASET_ID}.churn_retention_features`")
X = df.drop(columns=['churned'])
y = df['churned']

model = XGBClassifier()
model.fit(X, y)
model.predict(X).peek()

,predicted_churned,predicted_churned_probs,recency_days,frequency,monetary,avg_order_value,distinct_products,return_rate,age,gender,traffic_source,tenure_days
11729,True,"[{'label': True, 'prob': 0.9366042613983154} {'label': False, 'prob': 0.06339573860168457}]",206,1,0.02,0.02,1,0.0,46,F,Search,1664
2078,True,"[{'label': True, 'prob': 0.923842191696167} {'label': False, 'prob': 0.07615777850151062}]",1060,1,1.72,1.72,1,0.0,46,F,Organic,1299
19012,True,"[{'label': True, 'prob': 0.878886878490448} {'label': False, 'prob': 0.12111306935548782}]",333,1,1.82,1.82,1,0.0,57,F,Search,632
19645,True,"[{'label': True, 'prob': 0.9045749306678772} {'label': False, 'prob': 0.09542503952980042}]",80,1,1.82,1.82,1,0.0,52,F,Search,912
22067,True,"[{'label': True, 'prob': 0.9157683849334717} {'label': False, 'prob': 0.08423157036304474}]",1250,1,1.95,1.95,1,0.0,70,F,Search,1421


---
## Cleanup

Drop the models and feature table created in this notebook.

In [17]:
client.query(f"DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.churn_retention_baseline`").result()
print('Dropped churn_retention_baseline')

client.query(f"DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.churn_retention_model`").result()
print('Dropped churn_retention_model')

client.query(f"DROP TABLE IF EXISTS `{PROJECT_ID}.{DATASET_ID}.churn_retention_features`").result()
print('Dropped churn_retention_features')

Dropped churn_retention_baseline
Dropped churn_retention_model
Dropped churn_retention_features


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [18]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')